src/coreclr/jit/lowerxarch.cpp

```diff
@@ -3802,6 +3813,79 @@ GenTree* Lowering::TryLowerAndOpToResetLowestSetBit(GenTreeOp* andNode)
     return blsrNode;
 }
 
+//----------------------------------------------------------------------------------------------
+// Lowering::TryLowerAndOpToAndNot: Lowers a tree AND(X, NOT(Y)) to HWIntrinsic::AndNot
+//
+// Arguments:
+//    andNode - GT_AND node of integral type
+//
+// Return Value:
+//    Returns the replacement node if one is created else nullptr indicating no replacement
+//
+// Notes:
+//    Performs containment checks on the replacement node if one is created
+GenTree* Lowering::TryLowerAndOpToAndNot(GenTreeOp* andNode)
+{
+    assert(andNode->OperIs(GT_AND) && varTypeIsIntegral(andNode));
+
+    GenTree* opNode  = nullptr;
+    GenTree* notNode = nullptr;
+    if (andNode->gtGetOp1()->OperIs(GT_NOT))
+    {
+        notNode = andNode->gtGetOp1();
+        opNode  = andNode->gtGetOp2();
+    }
+    else if (andNode->gtGetOp2()->OperIs(GT_NOT))
+    {
+        notNode = andNode->gtGetOp2();
+        opNode  = andNode->gtGetOp1();
+    }
+
+    if (opNode == nullptr || !opNode->OperIs(GT_LCL_VAR))
```

Comment:
Why check for `opNode` being a local? Any arbitrary node should work.

Reply:
Because I knew that all the cases I was seeing hit this part were lcl_var. I can remove the check and see if anything breaks. Realistically are any other types likely to be seen here?

Reply:
It true that local nodes are the most common types of nodes (23% of all of them), but there is no reason to be restrictive here.

Reply:
Removing that checks increases the usage but introduces some new patterns that aren't improvements.
```       
not      edx
and      dword ptr [r8], edx
mov      rdx, rax
```
gets changed to:
```       
andn     edx, edx, dword ptr [r8]
mov      dword ptr [r8], edx
mov      rdx, rax
```
which is both longer and more instructions.

Reply:
It changes the overall results for libraries to be a regression. It might be best to keep the lcl_var check in place for now and later if someone has free time come back to it and see if there's a way to remove the check without regressions.

Reply:
> which is both longer and more instructions.

Should be about the same performance I suppose, but it is longer indeed. Should be ok to keep the check then, but please add a comment that it is a RMW-related heuristic, say:
```
We will avoid using "andn" when one of the operands is (likely to be) in memory: "and"'s RMW form provides for a shorter instruction sequence in that case.
```

Reply:
Moved to it's own check and comment added above it.

Reply:
> We will avoid using "andn" when one of the operands is (likely to be) in memory: "and"'s RMW form provides for a shorter instruction sequence in that case.

We have the following:
```
and  r/m32, r32             ; 1-byte opcode
and  r32,   r/m32           ; 1-byte opcode
andn r32a,  r32b, r/m32     ; 3-byte opcode
mov  r/m32, r32             ; 1-byte opcode
mov  r32,   r/m32           ; 1-byte opcode
not  r/m32                  ; 1-byte opcode
```

Where `n-byte opcode` is the "base encoding cost" of the instruction. Encoding the register or memory information is the same additional cost on top of that. And also considering the decoding, dispatching, and pipelining costs (`not, and` is two instructions and are dependent on eachother, etc).

In the "worst case", we have all three coming from memory (destination and both operands) and so the for `not, and` you have:
```
mov reg, [mem]
not reg
and reg, [mem]
mov [mem], reg
```
vs
```
mov reg, [mem]
andn reg, reg, [mem]
mov [mem], reg
```
so 3-bytes larger, but taking 1 less cycle

Basically every encodable scenario is the same except for "only the destination is memory" where you have:
```
not reg
and [mem], reg, reg
```
vs
```
not reg
andn reg, reg, reg
mov [mem], reg
```

and so having 1 less instruction and overall faster throughput

Reply:
If I change it to `if (IsBinOpInRMWStoreInd(andNode)) return nullptr;` then there a lot more improvements and the bad regression pattern seems to be gone and only size regressions remain. Does that seem like the right call?

Reply:
It's worth noting the result from the coreclr tests: 
`17346 total files with Code Size differences (17303 improved, 43 regressed)`

Reply:
The new check LGTM.